swayidle/main.c

```diff
@@ -0,0 +1,414 @@
+	char *endptr;
+	int seconds = strtoul(argv[1], &endptr, 10);
+	if (errno != 0 || *endptr != '\0') {
+		wlr_log(L_ERROR, "Invalid timeout parameter '%s', it should be a "
+				"numeric value representing seconds", optarg);
+		exit(-1);
+	}
+	struct swayidle_timeout_cmd *cmd =
+		calloc(1, sizeof(struct swayidle_timeout_cmd));
+	cmd->timeout = seconds * 1000;
+
+	wlr_log(L_DEBUG, "Register idle timeout at %d ms", cmd->timeout);
+	wlr_log(L_DEBUG, "Setup idle");
+	cmd->idle_cmd = parse_command(argc - 2, &argv[2]);
+
+	int result = 3;
+	if (argc >= 5 && !strcmp("resume", argv[3])) {
+		wlr_log(L_DEBUG, "Setup resume");
+		cmd->resume_cmd = parse_command(argc - 4, &argv[4]);
+		result = 5;
+	}
+	list_add(state.timeout_cmds, cmd);
+	return result;
+}
+
+int parse_sleep(int argc, char **argv) {
+	if (argc < 2) {
+		wlr_log(L_ERROR, "Too few parameters to before-sleep command. "
+				"Usage: before-sleep <command>");
+		exit(-1);
+	}
+
+	lock_cmd = parse_command(argc - 1, &argv[1]);
+	if (lock_cmd) {
+		wlr_log(L_DEBUG, "Setup sleep lock: %s", lock_cmd->param);
+	}
+
+	return 2;
+}
+
+
+int parse_args(int argc, char *argv[]) {
+	int c;
+
+	while ((c = getopt(argc, argv, "hs:d")) != -1) {
+		switch(c) {
+		case 'd':
+			debug = true;
+			break;
+		case 'h':
+		case '?':
+			printf("Usage: %s [OPTIONS]\n", argv[0]);
+			printf("  -d\tdebug\n");
+			printf("  -h\tthis help menu\n");
+			return 1;
+		default:
+			return 1;
+		}
+	}
+
+	if (debug) {
+		wlr_log_init(L_DEBUG, NULL);
+		wlr_log(L_DEBUG, "Loglevel debug");
+	} else {
+		wlr_log_init(L_INFO, NULL);
+	}
+
+
+	state.timeout_cmds = create_list();
+
+	int i = optind;
+	while (i < argc) {
+		if (!strcmp("timeout", argv[i])) {
+			wlr_log(L_DEBUG, "Got timeout");
+			i += parse_timeout(argc - i, &argv[i]);
+		} else if (!strcmp("before-sleep", argv[i])) {
+			wlr_log(L_DEBUG, "Got before-sleep");
+			i += parse_sleep(argc - i, &argv[i]);
+		} else {
+			wlr_log(L_ERROR, "Unsupported command '%s'", argv[i]);
+			exit(-1);
+		}
+	}
+	return 0;
+}
+
+void sway_terminate(int exit_code) {
+	if (state.event_loop) {
+		wl_event_loop_destroy(state.event_loop);
+	}
+	if (state.display) {
+		wl_display_disconnect(state.display);
+	}
+	exit(exit_code);
+}
+
+void sig_handler(int signal) {
+	sway_terminate(0);
+}
+
+static int display_event(int fd, uint32_t mask, void *data) {
+	wl_display_dispatch(state.display);
+	return 0;
+}
+
+void register_idle_timeout(void *item) {
+	struct swayidle_timeout_cmd *cmd = item;
+	if (cmd == NULL || !cmd->timeout) {
+		wlr_log(L_ERROR, "Invalid idle cmd, will not register");
+		return;
+	}
+	state.idle_timer =
+		org_kde_kwin_idle_get_idle_timeout(idle_manager, seat, cmd->timeout);
+	if (state.idle_timer != NULL) {
+		org_kde_kwin_idle_timeout_add_listener(state.idle_timer,
+				&idle_timer_listener, cmd);
+	} else {
+		wlr_log(L_ERROR, "Could not create idle timer");
+	}
+}
+
+int main(int argc, char *argv[]) {
+	signal(SIGINT, sig_handler);
+	signal(SIGTERM, sig_handler);
```

Comment:
Nitpick: is this really needed?

Reply:
Well, I guess it isn't striclty needed... but I think it is good to handle SIGINT and SIGTERM when you have a long running process, that handles shutdown in a controlled way.

Reply:
If the handler is just calling `exit(0)`, imho it can be removed

Reply:
Well, it destroys the eventloop and disconnects from the display... so it is not just exit 0... not sure how important it is to destroy the eventloop nicely and disconnect from the display... but, I don't see any reason not to do it.

Reply:
Ah, you're right, `sway_terminate` is not _just_ `exit(0)`.